### PR TITLE
implement objective envelopes (aka production envelopes)

### PR DESCRIPTION
**Pull request by exaexa**
## Description

There's the nice production envelopes that allow you to look at how much of something gets produced based on how much of certain metabolite goes in. This implements a slightly generalized version. There might be some extra features to implement, e.g. putting only unidirectional bounds of the optimizer.

#### Checklist

Still a draft.

- [x] code is formatted
- [x] Tests work (try with `]test COBREXA`)
- [x] All new functions have proper docstrings
- [x] Documentation builds correctly (try with `cd docs; julia make.jl`)

### CODE REVIEW

**Comment by cylon-x** [comment]
:red_square: &nbsp;Commit ab906d1e requires formatting!

Required formatting changes summary:
```
 src/analysis/envelopes.jl | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)
```

File: src/analysis/flux_variability_analysis.jl [hunk 1]
```diff
@@ -71,8 +71,8 @@ function flux_variability_analysis(
     # store a JuMP optimization model at all workers
     save_model = :(
         begin
-            model = $model
-            opt_model = $COBREXA.make_optimization_model(model, $optimizer)
+            local model = $model
+            local opt_model = $COBREXA.make_optimization_model(model, $optimizer)
```
Comment (exaexa) [inline src/analysis/flux_variability_analysis.jl:75]: todo: separate to own commit

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit aecb1233 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit aecb1233 is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 2502f82f is formatted properly.

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit fca020d2 is formatted properly.

File: src/analysis/screening.jl [hunk 2]
```diff
@@ -143,3 +143,80 @@ A shortcut for [`screen`](@ref) that only works with model variants.
...
+        modifications::Array{V,N},
+        analysis = screen_optimize_objective,
+        workers = [myid()],
+    ) where {V<:AbstractVector,N}
+
+Screen multiple modifications of the same optimization model.
+
+This function is potentially more efficient than [`screen`](@ref) because it
+avoids making variants of the model structure and remaking of the optimization
+model, but on the other hand can not express the scale of modifications.
```
Comment (stelmo) [inline src/analysis/screening.jl:169]: Could you be more clear here? I am not sure what this means

**Review by stelmo — APPROVED** [review]
This is awesome, but will definitely need a nice tutorial illustrating multiple aspects of these functions - they are so powerful that people might not use them because they don't understand their use cases. I would say that you'd need one tutorial for the phase plane/objective envelope stuff and one for the general purpose tutorial for `screen_optimize_objective`

File: src/analysis/screening.jl (hunk 2 above)
Comment (exaexa) [reply src/analysis/screening.jl]: boom.

**Comment by exaexa** [comment]
> This is awesome, but will definitely need a nice tutorial illustrating multiple aspects of these functions - they are so powerful that people might not use them because they don't understand their use cases.

well, opportunity to play with makie. :D

**Comment by cylon-x** [comment]
:green_circle: &nbsp;Commit 7dd04d4c is formatted properly.
